kernel/src/platform/mpu.rs

```diff
@@ -1,103 +1,109 @@
         Region {
-            base_address: (region_num as u32) | 1 << 4,
-            attributes: 0,
+            start: 0,
+            len: 0,
+            read: Permission::NoAccess,
+            write: Permission::NoAccess,
+            execute: Permission::NoAccess,
         }
     }
 
-    pub fn base_address(&self) -> u32 {
-        self.base_address
+    pub fn get_start(&self) -> usize {
+        self.start
+    }
+
+    pub fn get_len(&self) -> usize {
+        self.len
+    }
+
+    pub fn get_read_permission(&self) -> Permission {
+        self.read
+    }
+
+    pub fn get_write_permission(&self) -> Permission {
+        self.write
     }
 
-    pub fn attributes(&self) -> u32 {
-        self.attributes
+    pub fn get_execute_permission(&self) -> Permission {
+        self.execute
     }
 }
 
 pub trait MPU {
-    /// Enable the MPU.
-    ///
-    /// Both privileged and unprivileged code are subject to the constraints of
-    /// the active MPU regions. However, while unprivileged code cannot access
-    /// any memory space that is is not explicitly authorized to, privileged
-    /// code can access all unprotected (background) memory.
+    /// Enables the MPU.
     fn enable_mpu(&self);
 
-    /// Completely disable the MPU.
+    /// Disables the MPU.
     fn disable_mpu(&self);
 
-    /// Creates a new MPU-specific memory protection region
-    ///
-    /// `region_num`: an MPU region number 0-7
-    /// `start_addr`: the region base address. Lower bits will be masked
-    ///               according to the region size.
-    /// `len`       : region size as a PowerOfTwo (e.g. `16` for 64KB)
-    /// `execute`   : whether to enable code execution from this region
-    /// `ap`        : access permissions as defined in Table 4.47 of the user
-    ///               guide.
-    fn create_region(
-        region_num: usize,
-        start: usize,
-        len: usize,
-        execute: ExecutePermission,
-        access: AccessPermission,
-    ) -> Option<Region>;
+    /// Returns the number of supported MPU regions.
+    fn num_supported_regions(&self) -> u32;
 
-    /// Sets the base address, size and access attributes of the given MPU
-    /// region number.
-    fn set_mpu(&self, region: Region);
+    /// Allocates memory protection regions.
+    ///
+    /// # Arguments
+    ///
+    /// `regions`: array of regions to be allocated. The index of the array
+    ///            encodes the priority of the region. In the event of an
+    ///            overlap between regions, the implementor must ensure
+    ///            that the permissions of the region with higher priority
+    ///            take precendence.
+    ///
+    /// # Return Value
+    ///
+    /// If it is infeasible to allocate a memory region, returns its index
+    /// wrapped in a Result.
+    fn allocate_regions(&self, regions: &[Region]) -> Result<(), usize>;
 }
 
-/// Noop implementation of MPU trait
+/// No-op implementation of MPU trait
```

Comment:
I think this can be integrated into the trait directly (like how `Driver` is).

Reply:
Don't we need to implement it for `()` since there are chips that use that as their MPU?

Reply:
Is it possible for the chips to do `impl Mpu for ()` without defining any functions? I'm not sure if that would work in rust or not.

Reply:
There are maybe two questions:

1. Can you define an instance of a trait with methods without defining those methods: yes, if all methods have default implemetations.

2. Can the chip define an instance for unit (`()`): no. You can only define instances of traits in the module where the trait is defined, or the module where the type for which you want an instance is defined. Other forms are called orphan instances and they can create all sorts of safety issues so Rust doesn't allow them ATM (not just Rust, Haskell also disallowed them by default).